src/ucs/sys/topo.h

```diff
@@ -0,0 +1,100 @@
+/**
+* Copyright (C) NVIDIA Corporation. 2019.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifndef UCS_TOPO_H
+#define UCS_TOPO_H
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include <ucs/type/status.h>
+#include <limits.h>
+
+BEGIN_C_DECLS
+
+/** @file topo.h */
+
+/**
+ * @ingroup UCS_RESOURCE
+ * System Device abstraction
+ */
+typedef struct ucs_sys_device {
+    unsigned int id;             /**< Index of the unit */
+    unsigned int bus_id;         /**< bus ID of of the device if applicable*/
+    unsigned int numa_node;      /**< NUMA node assoicated with the device*/
+    char         name[PATH_MAX]; /**< Name of the device*/
+} ucs_sys_device_t;
+
+
+/**
+ * PCIe distance categories and answers what needs to be crossed to reach
+ */
+typedef enum {
+    UCS_DISTANCE_PIX = 0, /* traverse 1 PCIe switch */
+    UCS_DISTANCE_PXB,     /* traverse >1 PCIe switches */
+    UCS_DISTANCE_PHB,     /* traverse host bridge */
+    UCS_DISTANCE_NODE,    /* traverse host bridge in the same numa node */
+    UCS_DISTANCE_SYS      /* traverse CPU interconnect (like QPI) */
+} ucs_distance_t;
+
+
+/*
+ * Capture the estimated latency, bandwidth, and paths to traverse between two
+ * ucs_sys_device_t type structures
+ */
+typedef struct ucs_sys_dev_distance {
+    double latency;
+    double bandwidth;
+    ucs_distance_t distance;
+} ucs_sys_dev_distance_t;
+
+
+/**
+ * Find system device by pci bus id
+ *
+ * @param [in]     bus_id  bus id of the device of interest
+ * @param [in/out] sys_dev pointer to ucs_sys_device_t
+ *                         populated with device associated with the bus_id
+ *
+ * @return UCS_OK or error in case device cannot be found
+ */
+ucs_status_t ucs_topo_find_device_by_pci_id(int bus_id, ucs_sys_device_t *sys_dev);
```

Comment:
can cuda API return bus_id and pci_id?

Reply:
Yes. We can use:
```
CUresult cuDeviceGetPCIBusId ( char* pciBusId, int  len, CUdevice dev ) 
```

https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEM.html#group__CUDA__MEM_1g85295e7d9745ab8f0aa80dd1e172acfc

Reply:
i see pciBusId for cuda is char*, but in topo API it's an int. how does it work together?

Reply:
In https://github.com/openucx/ucx/pull/4321, I had a [helper function](https://github.com/openucx/ucx/pull/4321/files#diff-5951f5000a9ae730540b797fc8f7a779R96) to convert the char to an int.
Should I use char* instead of int?

Reply:
Changed to char* for now. Can revert if needed.

Reply:
so what's the difference between PCIe bus id and device name?

Reply:
For CUDA-capable GPUs, the intention was to get sys device by pci bus id. I was actually thinking that it makes sense to use `ibv_get_device_name` for `ucs_topo_find_device_by_name`.

Reply:
Making sure that I didn't misunderstand your question. Maybe you were checking for the actual difference in the output of cuDevice name and pci_bus calls. This program shows the diff:
```C
void print_device_name_bus_id(CUdevice dev_num)
{
    char name[PATH_MAX];
    char pciBusId[PATH_MAX];

    cuDeviceGetName(name, PATH_MAX, dev_num);
    printf("name of device %d = %s\n", (int) dev_num, name);
    cuDeviceGetPCIBusId (pciBusId, PATH_MAX, dev_num);
    printf("pci_bus_id of device %d = %s\n", (int) dev_num, pciBusId);
}

int main(int argc, char* argv[]) {

    cuInit(0);

    print_device_name_bus_id((CUdevice) 0);
    print_device_name_bus_id((CUdevice) 1);

    return 0;
}
```
Output on a dgx-1: (device name doesn't help differentiate devices)
```
name of device 0 = Tesla V100-SXM2-16GB
pci_bus_id of device 0 = 0000:06:00.0
name of device 1 = Tesla V100-SXM2-16GB
pci_bus_id of device 1 = 0000:07:00.0
```

Reply:
ok, so PCIe bus id is unique but device name is not unique.  so i think we should have API for pci bus ID only. i'm thinking if we should use it as "char*" or a "struct" with {domain,bus,slot,func} so it could be independent of string formatting.

Reply:
@yosefe 
The end goal is that we want sys_dev pointers so that we can have lanes per sys_devices (which have a memory subsystem - so not IB sys_devices). As this logic will likely be in `wireup/select.c`, I imagine we iterate through a list of global sys_device list, and for each sys_device with a memory subsystem we would want to find the sys_device associated with a transport instance that will be a candidate for the lane. For this, either we have to query the md properties for the transport instance for the device name; or we populate and query this above struct composed of {domain,bus,slot,func}. Once we have the sys_dev pointer, then we check distance and assign a score based on latency and b/w reported. Do I understand correctly?  If yes, then I can make the change. Just want to confirm before making changes. Lmk.

Reply:
An afterthought is that strcmp is probably faster than individually checking matches for domain, bus, slot, and function.

Reply:
@yosefe any further thoughts on this?